### PR TITLE
Reduce EDS Cluster memory utilization overhead from storing cluster name multiple times

### PR DESCRIPTION
**Pull request by yurykats**
Reduce EDS Cluster memory utilization overhead from storing cluster name multiple times.

This PR improves cluster memory efficiency in 2 ways:

1. It avoids `absl::optional` overhead when storing eds name in `ClusterInfoImpl` -- for non-EDS clusters `optional` overhead is 8 bytes, even though no eds name is used.
2. It avoids a copy of the eds cluster name in `EdsClusterImpl` -- since EdsClusterImpl holds a shared_ptr to ClusterInfoImpl already, and EdsClusterImpl holds the eds service name, EdsClusterImpl can get it from there, instead of keeping a copy.

This is a step towards more memory efficient config data structures. Issue #24154.

Commit Message:
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by stevenzzzz** [comment]
LGTM

File: source/common/upstream/upstream_impl.h
Comment (adisuissa) [inline source/common/upstream/upstream_impl.h:975]: While this may be correct, it seems to break the previous behavior (IIUC it returns nullopt in this case). This change may be ok, just wondering if it is warrented.

File: source/extensions/clusters/eds/eds.h
Comment (adisuissa) [inline source/extensions/clusters/eds/eds.h:91]: Can `cluster_name_` be kept as a reference to the `info_->name()` or `info_->edsServiceName()`?
(IIUC info is const and exists in the parent class so it should outlive the more specific class).

**Review by adisuissa — COMMENTED** [review]
Thanks, left a couple of comments.

File: source/common/upstream/upstream_impl.h
Comment (yurykats) [reply source/common/upstream/upstream_impl.h:975]: This allows us not to store `eds_name`, when it's the same as `name`, or when it's empty. Since `EdsClusterImpl` falls back on `name` in such cases already, I think this is a reasonable improvement.

File: source/extensions/clusters/eds/eds.h
Comment (yurykats) [reply source/extensions/clusters/eds/eds.h:91]: It could keep a reference to the optional `edsServiceName` (and `info->name()` returns a `string` referene), but I don't think there is much benefit to this, since it's not on a hot path.

File: source/common/upstream/upstream_impl.h
Comment (adisuissa) [reply source/common/upstream/upstream_impl.h:975]: Ok, hopefully it won't break anyone.
It should be noted that this has visible impact on clusters-dump in admin (see [here](https://github.com/envoyproxy/envoy/blob/4f2828650d7136e735f8f54e7fb0c9f74ad90ae7/source/server/admin/clusters_handler.cc#L118-L121) and [here](https://github.com/envoyproxy/envoy/blob/4f2828650d7136e735f8f54e7fb0c9f74ad90ae7/source/server/admin/clusters_handler.cc#L219-L222)), and on [load-stats report](https://github.com/envoyproxy/envoy/blob/4f2828650d7136e735f8f54e7fb0c9f74ad90ae7/source/common/upstream/load_stats_reporter.cc#L72-L74). So a release note should be added to this change.

Also please update the [unit tests](https://github.com/envoyproxy/envoy/blob/main/test/common/upstream/upstream_impl_test.cc) to cover the new behavior.

Comment (yurykats) [reply source/common/upstream/upstream_impl.h:975]: OK, if it has visible impact (I didn't realize that was the case), then it's not worth the trouble, I think. Especially since `EdsClusterImpl` falls back on cluster name anyway.

**Comment by phlax** [comment]
apologies - ive cancelled CI - please hold off pushing until we have progressed the releases that are currently going through

File: source/common/upstream/upstream_impl.h
Comment (yurykats) [reply source/common/upstream/upstream_impl.h]: Restored original behavior.

**Review by adisuissa — APPROVED** [review]
LGTM, thanks!

/assign-from @envoyproxy/senior-maintainers

**Review by yanavlasov — APPROVED** [review]

**Pull request merged**
